src/stan-users-guide/time-series.qmd

```diff
@@ -634,334 +634,116 @@ time of the original iterations.
-  real log_p_y_star;
-  {
-    array[T_unsup, K] int back_ptr;
-    array[T_unsup, K] real best_logp;
-    real best_total_logp;
-    for (k in 1:K) {
-      best_logp[1, k] = log(phi[k, u[1]]);
-    }
-    for (t in 2:T_unsup) {
-      for (k in 1:K) {
-        best_logp[t, k] = negative_infinity();
-        for (j in 1:K) {
-          real logp;
-          logp = best_logp[t - 1, j]
-                  + log(theta[j, k]) + log(phi[k, u[t]]);
-          if (logp > best_logp[t, k]) {
-            back_ptr[t, k] = j;
-            best_logp[t, k] = logp;
-          }
-        }
-      }
-    }
-    log_p_y_star = max(best_logp[T_unsup]);
-    for (k in 1:K) {
-      if (best_logp[T_unsup, k] == log_p_y_star) {
-        y_star[T_unsup] = k;
-      }
-    }
-    for (t in 1:(T_unsup - 1)) {
-      y_star[T_unsup - t] = back_ptr[T_unsup - t + 1,
-                                      y_star[T_unsup - t + 1]];
-    }
-  }
+  array[N] int latent_states = hmm_latent_rng(log_omega, gamma, rho);
+  matrix[3, N] hidden_probs = hmm_hidden_state_prob(log_omega, gamma, rho);
 }
 ```
-
-The bracketed block is used to make the three variables
-`back_ptr`, `best_logp`, and `best_total_logp`
-local so they will not be output.  The variable `y_star` will
-hold the label sequence with the highest probability given the input
-sequence `u`.  Unlike the forward algorithm, where the
-intermediate quantities were total probability, here they consist of
-the maximum probability `best_logp[t, k]` for the sequence up to
-time `t` with final output category `k` for time `t`,
-along with a backpointer to the source of the link.  Following the
-backpointers from the best final log probability for the final time
-`t` yields the optimal state sequence.
-
-This inference can be run for the same unsupervised outputs `u`
-as are used to fit the semisupervised model.  The above code can be
-found in the same model file as the unsupervised fit.  This is the
-Bayesian approach to inference, where the data being reasoned about is
-used in a semisupervised way to train the model.  It is not
-"cheating" because the underlying states for `u` are never
-observed --- they are just estimated along with all of the other
-parameters.
-
-If the outputs `u` are not used for semisupervised estimation but
-simply as the basis for prediction, the result is equivalent to what
-is represented in the BUGS modeling language via the cut operation.
-That is, the model is fit independently of `u`, then those
-parameters used to find the most likely state to have generated
-`u`.
-
+`hmm_hidden_state_prob` returns the marginal
+probabilities of each state, $\text{Pr}(z_n = k \mid \phi, y)$.
+This function cannot be used to compute the joint probability
+$\text{Pr}(z \mid \phi, y)$, because such calculation requires accounting
+for the posterior correlation between the different components of $z$.
+Therefore, `hidden_probs` should NOT be used to obtain posterior samples.
```

Comment:
NOT -> *not* (that'll make it italics, which is the standard way to add emphasis to typeset text).

This point is a bit more subtle, though.  You can use this to obtain posterior draws of the marginals $z_n \sim p(z_n \mid y)$, but if you put a sequence of draws from $p(z_1 \mid y), \ldots, p(z_N \mid y)$, you don't get a draw of $p(z \mid y)$.